runtime/runtime/src/lib.rs

```diff
@@ -1248,15 +1252,28 @@ impl Runtime {
             config.fees.fee(ActionCosts::new_action_receipt).exec_fee(),
         )?;
         let deposit_refund = if result.result.is_err() { total_deposit } else { 0 };
-        let gas_refund = if result.result.is_err() {
+        let gross_gas_refund = if result.result.is_err() {
             safe_add_gas(prepaid_gas, prepaid_exec_gas)? - result.gas_burnt
         } else {
             safe_add_gas(prepaid_gas, prepaid_exec_gas)? - result.gas_used
         };
+
+        // NEP-536 also adds a penalty to gas refund.
+        let refund_penalty: Gas = config.fees.gas_penalty_for_gas_refund(gross_gas_refund);
+        debug_assert!(
+            refund_penalty <= gross_gas_refund,
+            "gas_penalty_for_gas_refund returned larger penalty than input"
+        );
+        let net_gas_refund = gross_gas_refund - refund_penalty;
```

Comment:
Style:

```suggestion
        let net_gas_refund = gross_gas_refund.saturating_sub(refund_penalty);
```

or perhaps go all the way for 

```rust
let Some(net_gas_refund) = gross_gas_refund.checked_sub(refund_penalty) else {
    panic!("refund calculation botched")
};
```

Reply:
done